### PR TITLE
Mow, Plant and Harvest Tractor Blueprint

### PR DESCRIPTION
**Pull request by default-juice**
## Summary
Introduce the ability for a user to sign a blueprint that enables automated mowing planting and harvesting under certain conditions. 

### Smart Mow System
- **Buffer-based execution**: Only executes within 5 minutes of season end when protocol is about to print
- **Conditional mowing**: Requires minimum `twaDeltaB` (10e6) and claimable stalk (1 stalk) thresholds
- **Automatic mowing**: Triggers by default when planting or harvesting occurs

### Plant & Harvest Automation
- **Minimum thresholds**: Configurable minimums for plant (earned beans) and harvest (harvestable pods) amounts
- **Tip protection**: Enforces minimum amounts must exceed operator tip to prevent losses
- **Auto-compounding**: Harvested beans automatically deposited into Silo for immediate stalk generation

### Operator Management
- **Whitelisted operators**: Only approved operators can execute blueprints
- **Tip system**: Operators receive configurable Bean tips for execution
- **Withdrawal planning**: Automatic tip withdrawal from user's Silo positions

## Current Parameters

```solidity
struct MowPlantHarvestParams {
    uint256 minMowAmount;          // Minimum claimable stalk
    uint256 mintwaDeltaB;          // Minimum deltaB for smart mow
    uint256 minPlantAmount;        // Minimum earned beans to plant
    uint256 minHarvestAmount;      // Minimum harvestable pods
    uint8[] sourceTokenIndices;    // Tokens to withdraw tip from
    uint256 maxGrownStalkPerBdv;   // Max grown stalk per BDV of withdrawn deposits
    uint256 slippageRatio;         // LP withdrawal slippage
}
```

## Implementation Details

- **Season-based execution**: Prevents duplicate execution within same season
- **Comprehensive validation**: Checks user state, operator permissions, and parameter validity
- **Test coverage**: Full test suite covering all execution paths and edge cases

The blueprint enables yield optimization while maintaining user control through configurable parameters.

### CODE REVIEW

File: contracts/ecosystem/MowPlantHarvestBlueprint.sol [hunk 1]
```diff
@@ -0,0 +1,389 @@
...
+            LibTransfer.To.INTERNAL,
+            plan
+        );
+
+        // Tip the operator with the withdrawn beans
+        tractorHelpers.tip(
+            beanstalk.getBeanToken(),
+            vars.account,
+            vars.tipAddress,
+            params.opParams.operatorTipAmount,
```
Comment (fr1jo) [inline contracts/ecosystem/MowPlantHarvestBlueprint.sol:160]: i think we should have 3 (now 4) amounts for tipping, for mowing, planting, harvesting, and optimizing their plot deposits (see the comment related to how we figure out optimized plots). 

This is because the gas costs will vary on each user, so we should adjust accordingly. this can be abstracted from the user ofc

File: contracts/ecosystem/MowPlantHarvestBlueprint.sol [hunk 2]
```diff
@@ -0,0 +1,389 @@
...
+    ) internal view returns (bool) {
+        IBeanstalk.Season memory seasonInfo = beanstalk.time();
+
+        // if the time until next season is more than the buffer don't mow, too early
+        uint256 nextSeasonExpectedTimestamp = seasonInfo.timestamp + seasonInfo.period;
+        if (nextSeasonExpectedTimestamp - block.timestamp > SMART_MOW_BUFFER) return false;
+
+        // if the totalDeltaB and totalClaimableStalk are both greater than the min amount, return true
+        // This also guards against double dipping the blueprint after planting or harvesting since stalk will be 0
+        return beanstalk.totalDeltaB() > int256(mintwaDeltaB) && totalClaimableStalk > minMowAmount;
```
Comment (fr1jo) [inline contracts/ecosystem/MowPlantHarvestBlueprint.sol:259]: this line should be reversed because if `totalClaimableStalk > minMowAmount` is false, then the other statement in the AND will not be called. This is important because beanstalk.totalDeltaB() takes in a lot of information and gas.

File: contracts/ecosystem/MowPlantHarvestBlueprint.sol [hunk 3]
```diff
@@ -0,0 +1,389 @@
...
+                "Min harvest amount must be greater than operator tip amount"
+            );
+        }
+    }
+
+    /// @dev validates info related to the blueprint such as the order hash and the last executed season
+    function _validateBlueprint(bytes32 orderHash) internal view {
+        require(orderHash != bytes32(0), "No active blueprint, function must run from Tractor");
+        require(
+            orderLastExecutedSeason[orderHash] < beanstalk.time().current,
```
Comment (fr1jo) [inline contracts/ecosystem/MowPlantHarvestBlueprint.sol:380]: beanstalk.time() is called multiple times in this file. it should be cached and used instead of recalling.

File: contracts/ecosystem/MowPlantHarvestBlueprint.sol (hunk 2 above)
Comment (default-juice) [reply contracts/ecosystem/MowPlantHarvestBlueprint.sol]: Addressed here [ef68876](https://github.com/pinto-org/protocol/pull/134/commits/ef688767350cde476cc9ebdd356809095fd27948)

File: contracts/ecosystem/MowPlantHarvestBlueprint.sol [hunk 4]
```diff
@@ -0,0 +1,389 @@
...
+        // Execute operations in order: mow first (if needed), then plant, then harvest
+        if (vars.shouldMow) beanstalk.mowAll(vars.account);
+
+        // Plant if the conditions are met
+        if (vars.shouldPlant) beanstalk.plant();
+
+        // Harvest if the conditions are met
+        if (vars.shouldHarvest) {
+            uint256 harvestedBeans = beanstalk.harvest(
+                beanstalk.activeField(),
```
[the remark this answers is not in the document]
Comment (default-juice) [reply contracts/ecosystem/MowPlantHarvestBlueprint.sol:182]: Since we keep in mind gas cost in this, i believe its safe to store the active field and the bean token as constants here instead of making calls to the diamond. The likelihood of those values changing is very low.

File: contracts/ecosystem/MowPlantHarvestBlueprint.sol [hunk 5]
```diff
@@ -0,0 +1,389 @@
...
+        uint256 totalHarvestablePods;
+        bool shouldMow;
+        bool shouldPlant;
+        bool shouldHarvest;
+        uint256[] harvestablePlots;
+        LibTractorHelpers.WithdrawalPlan plan;
+    }
+
+    // Mapping to track the last executed season for each order hash
+    mapping(bytes32 orderHash => uint32 lastExecutedSeason) public orderLastExecutedSeason;
```
[the remark this answers is not in the document]
Comment (default-juice) [reply contracts/ecosystem/MowPlantHarvestBlueprint.sol:87]: Agree on this. Generally, i tried to keep styling in line with the `SowBlueprint` given it was thoroughly reviewed and audited but open to refactoring the structure moving forward. (same comment as below)

File: contracts/ecosystem/MowPlantHarvestBlueprint.sol [hunk 6]
```diff
@@ -0,0 +1,389 @@
...
+     */
+    struct MowPlantHarvestBlueprintStruct {
+        MowPlantHarvestParams mowPlantHarvestParams;
+        OperatorParams opParams;
+    }
+
+    /**
+     * @notice Struct to hold mow, plant and harvest parameters
+     * @param minMowAmount The minimum total claimable stalk threshold to mow
+     * @param mintwaDeltaB The minimum twaDeltaB to mow if the protocol is close to printing
```
[the remark this answers is not in the document]
Comment (default-juice) [reply contracts/ecosystem/MowPlantHarvestBlueprint.sol]: Addressed here [a52eac5](https://github.com/pinto-org/protocol/pull/134/commits/a52eac5f484b6881968ee6c17f7646f4cf9108d8)

File: contracts/ecosystem/MowPlantHarvestBlueprint.sol [hunk 7]
```diff
@@ -0,0 +1,389 @@
...
+        uint256 slippageRatio;
+    }
+
+    /**
+     * @notice Struct to hold operator parameters
+     * @param whitelistedOperators What operators are allowed to execute the blueprint
+     * @param tipAddress Address to send tip to
+     * @param operatorTipAmount Amount of tip to pay to operator
+     */
+    struct OperatorParams {
```
[the remark this answers is not in the document]
Comment (default-juice) [reply contracts/ecosystem/MowPlantHarvestBlueprint.sol:64]: Let's try to abstract blueprints as much as possible. Started work on this draft PR an open to comments. Im thinking we identify the common validations that bluerpints should do and add virtual functions in the base contract that blueprint implementations will override: #141

File: contracts/ecosystem/MowPlantHarvestBlueprint.sol (hunk 5 above)
[the remark this answers is not in the document]
Comment (default-juice) [reply contracts/ecosystem/MowPlantHarvestBlueprint.sol:90]: See comment above

File: contracts/ecosystem/MowPlantHarvestBlueprint.sol [hunk 8]
```diff
@@ -0,0 +1,389 @@
...
+import {PerFunctionPausable} from "./PerFunctionPausable.sol";
+import {LibTractorHelpers} from "contracts/libraries/Silo/LibTractorHelpers.sol";
+
+/**
+ * @title MowPlantHarvestBlueprint
+ * @author DefaultJuice
+ * @notice Contract for mowing, planting and harvesting with Tractor, with a number of conditions
+ */
+contract MowPlantHarvestBlueprint is PerFunctionPausable {
+    /// @dev Buffer for operators to check if the protocol is close to printing
```
[the remark this answers is not in the document]
Comment (fr1jo) [reply contracts/ecosystem/MowPlantHarvestBlueprint.sol]: lgtm!

File: contracts/ecosystem/MowPlantHarvestBlueprint.sol (hunk 5 above)
[the remark this answers is not in the document]
Comment (fr1jo) [reply contracts/ecosystem/MowPlantHarvestBlueprint.sol:90]: great work

File: contracts/ecosystem/MowPlantHarvestBlueprint.sol [hunk 9]
```diff
@@ -0,0 +1,389 @@
...
+                count++;
+            }
+        }
+
+        // Allocate the array
+        harvestablePlots = new uint256[](count);
+        uint256 j = 0;
+
+        // Now, fill the array and sum pods
+        for (uint256 i = 0; i < plots.length; i++) {
```
[the remark this answers is not in the document]
Comment (fr1jo) [reply contracts/ecosystem/MowPlantHarvestBlueprint.sol]: re. 3) - when we fetch a `Plot`, since the data storage is 2 slots long, reaching each plot would take 10k to read. If we only fetch the index in a Plot, then reading each index would only take 5k gas for a first read instead. 
```
 struct Plot {
        uint256 index;
        uint256 pods;
    }
```
"Updating the function such that it breaks out the moment that the harvestable plot exceeds the min harvest also dramatically decreases the cost." - good call, we should harvest on all their pods, not just the min amount.

File: contracts/ecosystem/MowPlantHarvestBlueprint.sol (hunk 4 above)
[the remark this answers is not in the document]
Comment (fr1jo) [reply contracts/ecosystem/MowPlantHarvestBlueprint.sol:185]: see comment above

File: contracts/ecosystem/MowPlantHarvestBlueprint.sol (hunk 2 above)
Comment (fr1jo) [reply contracts/ecosystem/MowPlantHarvestBlueprint.sol]: lgtm

File: contracts/beanstalk/facets/field/FieldFacet.sol [hunk 10]
```diff
@@ -452,4 +477,122 @@ contract FieldFacet is Invariable, ReentrancyGuard {
...
         }
     }
+
+    //////////////////// PLOT INDEX HELPERS ////////////////////
+
+    /**
+     * @notice returns Plot indexes by their positions in the plotIndexes array.
+     * @dev plotIndexes is an array of Plot indexes, used to return the farm plots of a Farmer.
+     */
+    // 3. A function that, given a farmer and index of an array, returns only the 'index' portion of the plot.
```
Comment (fr1jo) [inline contracts/beanstalk/facets/field/FieldFacet.sol:487]: can we remove the 3. and clean this up?

File: contracts/beanstalk/facets/field/FieldFacet.sol [hunk 11]
```diff
@@ -452,4 +477,122 @@ contract FieldFacet is Invariable, ReentrancyGuard {
...
+    /**
+     * @notice Combines an account's adjacent plots.
+     * @param account The account that owns the plots to combine
+     * @param fieldId The field ID containing the plots
+     * @param plotIndexes Array of adjacent plot indexes to combine (must be sorted and consecutive)
+     * @dev Plots must be adjacent: plot[i].index + plot[i].pods == plot[i+1].index
+     *      Updates storage by merging plots and rebuilding plotIndexes/piIndex
+     *      Any account can combine any other account's adjacent plots
+     */
+    function combinePlots(
```
Comment (fr1jo) [inline contracts/beanstalk/facets/field/FieldFacet.sol:533]: in this function, you combine the plots together, and then have a seperate function that adjusts `plotIndexes` and `piIndex` as well. 

given that you already know the index you're removing in `combinePlots`, we can directly delete the entry in `piIndex` below the line where we delete the plot (i.e `delete s.accts[account].fields[fieldId].plots[plotIndexes[i]];`)

Comment (fr1jo) [reply contracts/beanstalk/facets/field/FieldFacet.sol:533]: once you do that, we should just call  `LibDibbler.removePlotIndexFromAccount(account, fieldId, index);`

File: contracts/beanstalk/facets/field/FieldFacet.sol [hunk 12]
```diff
@@ -16,6 +16,7 @@ import {LibDibbler} from "contracts/libraries/LibDibbler.sol";
 import {LibDiamond} from "contracts/libraries/LibDiamond.sol";
 import {LibMarket} from "contracts/libraries/LibMarket.sol";
 import {BeanstalkERC20} from "contracts/tokens/ERC20/BeanstalkERC20.sol";
+import "forge-std/console.sol";
```
Comment (fr1jo) [inline contracts/beanstalk/facets/field/FieldFacet.sol:19]: pls remove

**Review by fr1jo — COMMENTED** [review]
review II

File: contracts/beanstalk/facets/field/FieldFacet.sol [hunk 13]
```diff
@@ -452,4 +474,83 @@ contract FieldFacet is Invariable, ReentrancyGuard {
...
+
+            // check adjacency: expected next start == current plot start
+            require(expectedNextStart == plotIndexes[i], "Field: Plots to combine not adjacent");
+
+            totalPods += currentPods;
+            expectedNextStart = plotIndexes[i] + currentPods;
+
+            // delete subsequent plot, plotIndex and piIndex mapping entry
+            delete s.accts[account].fields[fieldId].plots[plotIndexes[i]];
+            LibDibbler.removePlotIndexFromAccount(account, fieldId, plotIndexes[i]);
```
[the remark this answers is not in the document]
Comment (default-juice) [reply contracts/beanstalk/facets/field/FieldFacet.sol:550]: But generally this seems to introduce complexity which i'm not sure is worth having if the gas cost increases by ~n*5000 for 1 transaction periodically ( 0,000000000007369547 (eth per gas) * 100 (plots) * 5000 ( gas of sstore from non zero to non zero value) * 4000 (current eth price) ~= 0,014 cents once every 100 plots for example ( a total of 1,24 usd annualized if this is the only account that sows every hour.

File: contracts/beanstalk/facets/field/FieldFacet.sol [hunk 14]
```diff
@@ -452,4 +474,83 @@ contract FieldFacet is Invariable, ReentrancyGuard {
...
+        }
+
+        // update first plot with combined pods
+        s.accts[account].fields[fieldId].plots[plotIndexes[0]] = totalPods;
+    }
```
[the remark this answers is not in the document]
Comment (default-juice) [reply contracts/beanstalk/facets/field/FieldFacet.sol:565]: Added an event here with the requested fields. [f2fd59f](https://github.com/pinto-org/protocol/pull/134/commits/f2fd59f9e5557bf13bb3063080632d861d455ba1). @PintoPirate Lmk if you need anything else

File: contracts/ecosystem/MowPlantHarvestBlueprint.sol (hunk 1 above)
Comment (default-juice) [reply contracts/ecosystem/MowPlantHarvestBlueprint.sol]: What do you think about putting the optimizations in a separate blueprint that the user signs once and handles both field (plot) optimization and deposit optimization after some configured thresholds? 

Added separate tips for mowing planting and harvesting here [63e8113](https://github.com/pinto-org/protocol/pull/134/commits/63e8113c27bc0ef5357b666ccff69c543e16de60) . Tests need to be updated but stack too deep makes it pretty annoying

File: contracts/beanstalk/facets/field/FieldFacet.sol [hunk 15]
```diff
@@ -553,46 +545,12 @@ contract FieldFacet is Invariable, ReentrancyGuard {
...
             expectedNextStart = plotIndexes[i] + currentPods;
 
-            // delete subsequent plots, set the amount to 0 so that we can rebuild the array later
+            // delete subsequent plot, plotIndex and piIndex mapping entry
             delete s.accts[account].fields[fieldId].plots[plotIndexes[i]];
+            LibDibbler.removePlotIndexFromAccount(account, fieldId, plotIndexes[i]);
         }
 
         // update first plot with combined pods
         s.accts[account].fields[fieldId].plots[plotIndexes[0]] = totalPods;
```
[the remark this answers is not in the document]
Comment (default-juice) [reply contracts/beanstalk/facets/field/FieldFacet.sol:568]: > 1. can you move the tests related to combining plots to the field tests? why is it in the mow plant harvest blueprint?
> 2. can you have a test that attempts to combine a plot like the example above?

Im in the process of refactoring and cleaning up the tests, including this.

[the remark this answers is not in the document]
Comment (default-juice) [reply contracts/beanstalk/facets/field/FieldFacet.sol:568]: Cleaned up tests here: [57d0be0](https://github.com/pinto-org/protocol/pull/134/commits/57d0be02e4c266ab89eacedadc695ba7f47d7874)

File: contracts/ecosystem/MowPlantHarvestBlueprint.sol (hunk 1 above)
Comment (default-juice) [reply contracts/ecosystem/MowPlantHarvestBlueprint.sol]: Fixed the tests here [57d0be0](https://github.com/pinto-org/protocol/pull/134/commits/57d0be02e4c266ab89eacedadc695ba7f47d7874)

File: contracts/ecosystem/MowPlantHarvestBlueprint.sol [hunk 16]
```diff
@@ -0,0 +1,355 @@
...
+     * Only applicable for lp token withdrawals.
+     */
+    struct MowPlantHarvestParams {
+        // Mow
+        uint256 minMowAmount;
+        uint256 mintwaDeltaB;
+        // Plant
+        uint256 minPlantAmount;
+        // Harvest
+        uint256 minHarvestAmount;
```
[the remark this answers is not in the document]
Comment (default-juice) [reply contracts/ecosystem/MowPlantHarvestBlueprint.sol:50]: Are you referring to having something an array of configs for each fieldId in the params instead ? 

```solidity
    /**
     * @notice Struct to hold field-specific harvest configuration
     * @param fieldId The field ID to harvest from
     * @param minHarvestAmount The minimum harvestable pods threshold for this field
     */
    struct FieldHarvestConfig {
        uint256 fieldId;
        uint256 minHarvestAmount;
    }
    
        /**
     * @notice Struct to hold mow, plant and harvest parameters
     * @param minMowAmount The minimum total claimable stalk threshold to mow
     * @param mintwaDeltaB The minimum twaDeltaB to mow if the protocol
     * is close to starting the next season above the value target
     * @param minPlantAmount The earned beans threshold to plant
     * @param fieldHarvestConfigs Array of field-specific harvest configurations
     * -----------------------------------------------------------
     * @param sourceTokenIndices Indices of source tokens to withdraw from
     * @param maxGrownStalkPerBdv Maximum grown stalk per BDV allowed
     * @param slippageRatio The price slippage ratio for lp token withdrawal.
     * Only applicable for lp token withdrawals.
     */
    struct MowPlantHarvestParams {
        // Mow
        uint256 minMowAmount;
        uint256 mintwaDeltaB;
        // Plant
        uint256 minPlantAmount;
        // Harvest, per field id
 @>  FieldHarvestConfig[] fieldHarvestConfigs;
        // Withdrawal plan parameters for tipping
        uint8[] sourceTokenIndices;
        uint256 maxGrownStalkPerBdv;
        uint256 slippageRatio;
    }
    
    ....
 ```

File: contracts/beanstalk/facets/field/FieldFacet.sol (hunk 15 above)
[the remark this answers is not in the document]
Comment (fr1jo) [reply contracts/beanstalk/facets/field/FieldFacet.sol:568]: Looking at the tests, there isn't a test where we check that the final plot index that's being merged into is not the one at the earliest index. We need a test that tries to combine the plots in the example above. Even if we don't think that this behavior would break from our understanding, it is good and standard practice to test these edge cases. 

you can do this by doing multiple sets of sows like you do in `setUpMultipleConsecutiveAccountPlots`, then send the earliest plot to another account and back. this will put the earliest index at the end. 
ex: given someones index of [1000, 2000, 3000] -> send plot index 1000 to someone else -> index is now [2000, 3000], user receives index 1000 -> index is now [2000, 3000, 1000].
